hal/src/stm32f2xx/i2c_hal.c

```diff
@@ -1,726 +1,740 @@
+{
+    uint32_t _millis;
+    uint8_t bytesRead = 0;
+
+    // clamp to buffer length
+    if(quantity > BUFFER_LENGTH)
+    {
+        quantity = BUFFER_LENGTH;
+    }
+
+    /* Send START condition */
+    I2C_GenerateSTART(i2cMap[i2c]->I2C_Peripheral, ENABLE);
+
+    _millis = HAL_Timer_Get_Milli_Seconds();
+    while(!I2C_CheckEvent(i2cMap[i2c]->I2C_Peripheral, I2C_EVENT_MASTER_MODE_SELECT))
+    {
+        if(EVENT_TIMEOUT < (HAL_Timer_Get_Milli_Seconds() - _millis))
+        {
+            /* SW Reset the I2C Peripheral */
+            HAL_I2C_SoftwareReset(i2c);
+
+            return 0;
+        }
+    }
+
+    /* perform blocking read into buffer */
+    uint8_t *pBuffer = i2cMap[i2c]->rxBuffer;
+    uint8_t numByteToRead = quantity;
+
+	if (quantity == 1)
+		/* Disable Acknowledgement */
+		I2C_AcknowledgeConfig(i2cMap[i2c]->I2C_Peripheral, DISABLE);
+	else
+		I2C_AcknowledgeConfig(i2cMap[i2c]->I2C_Peripheral, ENABLE);
+
+    /* Send Slave address for read */
+    I2C_Send7bitAddress(i2cMap[i2c]->I2C_Peripheral, address << 1, I2C_Direction_Receiver);
+
+    _millis = HAL_Timer_Get_Milli_Seconds();
+    while(!I2C_CheckEvent(i2cMap[i2c]->I2C_Peripheral, I2C_EVENT_MASTER_RECEIVER_MODE_SELECTED))
+    {
+        if(EVENT_TIMEOUT < (HAL_Timer_Get_Milli_Seconds() - _millis))
+        {
+            /* Send STOP Condition */
+            I2C_GenerateSTOP(i2cMap[i2c]->I2C_Peripheral, ENABLE);
+
+            /* SW Reset the I2C Peripheral */
+            HAL_I2C_SoftwareReset(i2c);
+
+            return 0;
+        }
+    }
+
+
+    /* While there is data to be read */
+    while(numByteToRead)
+    {
+        /* Wait for data register to be full */
+		/* This means the byte in the RX has already been acked */
+		/* Ack is sent on the xfer of teh byte to RX from the shift register */
+        _millis = HAL_Timer_Get_Milli_Seconds();
+        while(I2C_GetFlagStatus(i2cMap[i2c]->I2C_Peripheral, I2C_FLAG_RXNE) == RESET)
+        {
+            if(EVENT_TIMEOUT < (HAL_Timer_Get_Milli_Seconds() - _millis))
+            {
+                /* SW Reset the I2C Peripheral */
+                HAL_I2C_SoftwareReset(i2c);
+
+                return 0;
+            }
+        }
+
+        switch	(numByteToRead) {
+            case 2:
+                /* Disable Acknowledgement on last byte which is being assembled in shift register*/
+                I2C_AcknowledgeConfig(i2cMap[i2c]->I2C_Peripheral, DISABLE);
+                break;
+
+            case 1:
+                /* Send STOP Condition */
+                if (stop == true)
+                        I2C_GenerateSTOP(i2cMap[i2c]->I2C_Peripheral, ENABLE);
+                break;
+
+            default:
+                break;
+        }
+
+        /* Read the byte from the Slave */
+		/* This will enhable the transfer of the next byte (if any) from the shift register to the RX */
+		/* and also sending ACK/NACK for that byte */

```

Comment:
These comments are easier to understand, but some reference to numByteToRead >2, == 2 and == 1 in context would be helpful.